### PR TITLE
Compiler refactor

### PR DESCRIPTION
**Pull request by Kodiologist**
More could be done, but this is progress. It saves a few hundred lines from the compiler by removing various kinds of duplication. There should be no changes to Hy semantics, although the exact generated AST is different in some cases.

### CODE REVIEW

**Comment by Kodiologist** [comment]
Huh, it looks like the redundant `arg` argument to some nodes is required by Python 3.3 but not later versions nor Python 2. [Dumb old 3.3 is going to officially go out of its support window in a month anyway.](https://docs.python.org/devguide/#status-of-python-branches) Do you guys think it's okay if we just drop 3.3 now? If I fix this, we'll just have to remove the fix again in October.

**Comment by gilch** [comment]
Make a separate PR to drop 3.3, and we'll discuss it there. I'm okay with it, since it's only a month away. I doubt anyone will care, but they can install Hy from an older commit for a month if they have to. I don't think the date is final, but changes to that seem improbable at this point.

**Comment by Kodiologist** [comment]
> Make a separate PR to drop 3.3, and we'll discuss it there.

Okay, it's #1401.

**Comment by Kodiologist** [comment]
Rebased for the merge of that PR.

File: hy/compiler.py [hunk 1]
```diff
@@ -2521,52 +2156,33 @@ def compile_symbol(self, symbol):
...
-    @builds(HyKeyword)
-    def compile_keyword(self, keyword):
-        return ast.Str(s=str_type(keyword),
-                       lineno=keyword.start_line,
-                       col_offset=keyword.start_column)
+        return asty.Name(symbol, id=ast_str(symbol), ctx=ast.Load())
+
+    @builds(HyString, HyKeyword, HyBytes)
+    def compile_string(self, string, building):
+        node = asty.Bytes if PY3 and building is HyBytes and PY3 else asty.Str
```
Comment (gilch) [inline hy/compiler.py]: `PY3 and building is HyBytes and PY3` checking it twice?

Comment (Kodiologist) [reply hy/compiler.py]: Oops, I must've been in a Christmasy mood ("He's makin' a list, he's checkin' it twice…"). Fixed.

File: tests/native_tests/language.hy [hunk 2]
```diff
@@ -1592,3 +1592,14 @@
...
+#@(pytest.mark.xfail
+(defn test-assert-multistatements []
+  ; https://github.com/hylang/hy/issues/1390
+  (setv l (set))
+  (defn f [x]
+    (.add l x)
+    True)
+  (assert (do (f 1) (f 2)) (do (f 3) (f 4)))
+  (print l)
+  (assert (= l [1 2 3 4]))))
```
Comment (gilch) [inline tests/native_tests/language.hy]: This seems unrelated to the rest of the PR.

File: hy/compiler.py [hunk 3]
```diff
@@ -1125,117 +1080,60 @@ def compile_if(self, expression):
...
-                          lineno=expr.start_line,
-                          col_offset=expr.start_column)
-
     @builds("nonlocal")
     @checkargs(min=1)
-    def compile_nonlocal_expression(self, expr):
-        if not PY3:
+    def compile_global_or_nonlocal(self, expr):
+        form = expr.pop(0)
+        if form == "nonlocal" and not PY3:
```
Comment (gilch) [inline hy/compiler.py]: shouldn't this be in the `@builds_if` decorator?

File: tests/native_tests/language.hy (hunk 2 above)
Comment (Kodiologist) [reply tests/native_tests/language.hy]: It is. I put that commit in because I noticed the bug while reading through the compiler. I figured that adding an xfail test wasn't worth its own PR.

File: hy/compiler.py (hunk 3 above)
Comment (Kodiologist) [reply hy/compiler.py]: Yeah, good idea.

Comment (Kodiologist) [reply hy/compiler.py]: Okay, changed and rebased.

**Comment by gilch** [comment]
@hylang/core can we get another ack? Any objections?

**Comment by kirbyfan64** [comment]
I'd review, but this is a huge PR, and I have neither the time nor energy right now to go through it, and it'd be silly for me to approve for the sake of approving...

**Comment by Kodiologist** [comment]
@kirbyfan64 The commit "General compiler refactoring pass" has most of the change in terms of sheer line count, but not much interesting is going on there. You could just review this PR on the basis of the other commits, if you like.

**Comment by gilch** [comment]
@Kodiologist, there's a conflict due to #1418. #1419 might cause similar problems.

**Comment by kirbyfan64** [comment]
hold on, manually testing and merging...
